x/batch/keeper/delegation.go

```diff
@@ -0,0 +1,130 @@
+	rewardsTotal := sdk.Coins{}
+	remainderTotal := sdk.DecCoins{}
+	k.stakingKeeper.IterateDelegations(ctx, delAddr, func(_ int64, del stakingtypes.DelegationI) (stop bool) {
+		valAddr := del.GetValidatorAddr()
+		val := k.stakingKeeper.Validator(ctx, valAddr)
+
+		// check existence of delegator starting info
+		if !k.distrKeeper.HasDelegatorStartingInfo(ctx, del.GetValidatorAddr(), del.GetDelegatorAddr()) {
+			return false
+		}
+
+		// end current period and calculate rewards
+		endingPeriod := k.distrKeeper.IncrementValidatorPeriod(ctx, val)
+		rewardsRaw := k.distrKeeper.CalculateDelegationRewards(ctx, val, del, endingPeriod)
+		outstanding := k.distrKeeper.GetValidatorOutstandingRewardsCoins(ctx, del.GetValidatorAddr())
+
+		// defensive edge case may happen on the very final digits
+		// of the decCoins due to operation order of the distribution mechanism.
+		rewards := rewardsRaw.Intersect(outstanding)
+		if !rewards.IsEqual(rewardsRaw) {
+			logger := k.Logger(ctx)
+			logger.Info(
+				"rounding error withdrawing rewards from validator",
+				"delegator", del.GetDelegatorAddr().String(),
+				"validator", val.GetOperator().String(),
+				"got", rewards.String(),
+				"expected", rewardsRaw.String(),
+			)
+		}
+
+		// truncate reward dec coins, return remainder to community pool
+		finalRewards, remainder := rewards.TruncateDecimal()
+		remainderTotal = remainderTotal.Add(remainder...)
+		rewardsTotal = rewardsTotal.Add(finalRewards...)
+
+		// update the outstanding rewards and the community pool only if the
+		// transaction was successful
+		k.distrKeeper.SetValidatorOutstandingRewards(ctx, del.GetValidatorAddr(), types.ValidatorOutstandingRewards{Rewards: outstanding.Sub(rewards)})
+
+		// decrement reference count of starting period
+		startingInfo := k.distrKeeper.GetDelegatorStartingInfo(ctx, del.GetValidatorAddr(), del.GetDelegatorAddr())
+		startingPeriod := startingInfo.PreviousPeriod
+		k.decrementReferenceCount(ctx, del.GetValidatorAddr(), startingPeriod)
+
+		// reinitialize the delegation
+		k.initializeDelegation(ctx, valAddr, delAddr)
+
+		ctx.EventManager().EmitEvent(
+			sdk.NewEvent(
+				types.EventTypeWithdrawRewards,
+				sdk.NewAttribute(sdk.AttributeKeyAmount, finalRewards.String()),
+				sdk.NewAttribute(types.AttributeKeyValidator, val.GetOperator().String()),
+			),
+		)
+		return false
+	})
+
+	// distribute total remainder to community pool
+	feePool := k.distrKeeper.GetFeePool(ctx)
+	feePool.CommunityPool = feePool.CommunityPool.Add(remainderTotal...)
+	k.distrKeeper.SetFeePool(ctx, feePool)
+
+	// add total reward coins to user account
+	if !rewardsTotal.IsZero() {
+		withdrawAddr := k.distrKeeper.GetDelegatorWithdrawAddr(ctx, delAddr)
+		err := k.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, withdrawAddr, rewardsTotal)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		baseDenom, _ := sdk.GetBaseDenom()
+		rewardsTotal = sdk.Coins{sdk.Coin{
```

Comment:
can this just be an empty `sdk.Coins{}`?

Reply:
sure